### ISSUE TITLE
Pay attention to your Blacklist URLs - they get back some unset headers

### ISSUE THREAD

**Issue by ADDISON74** [nexcess/magento-turpentine#1055]
I discovered many URLs in Magento Default which are cached. Here is one example "Popular Search Terms" page: http://www.mydomain.com/catalogsearch/term/popular/

If you delete your search terms in backend then reload "Popular Search Terms" page all search terms are still there. Once you restart your Varnish service and reload the page there are no search words.

I added catalogsearch/term/popular to URL Blacklist in Backend. A good way to check if this link is not cached is looking at time load. Click two times the link and see if there are any notable differences. In my case there are no changes. Page is loaded from Apache.

Now comes the nice part. If I investigate the headers for this page, I am surprised to find out headers like "X-Powered-By", "Server" are showing up. This is not good at all. By default Varnish is hiding this information inside VCL file, by the end of the file: 

```php
unset resp.http.X-Varnish;
unset resp.http.Via;
unset resp.http.X-Powered-By;
unset resp.http.Server;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;
```

How can I keep some headers hidden if the page is not cached (in URL Blacklist)? If there any part inside VCL, probably an if statement where to unset some hearders (if Blacklist unset these headers)?

**Comment by aricwatson** [comment]
> I added catalogsearch/term/popular to URL Blacklist in Backend. A good way to check if this link is not cached is looking at time load. Click two times the link and see if there are any notable differences. In my case there are no changes. Page is loaded from Apache.

Alternatively, you can reload a couple times then look at the page source to see if there are any ESI block html comments present.

> How can I keep some headers hidden if the page is not cached (in URL Blacklist)? If there any part inside VCL, probably an if statement where to unset some hearders (if Blacklist unset these headers)?

If a URL is matches a blacklist entry, then it bypasses Varnish - so nothing in the current VCL will affect them. 

However, I haven't tested this (yet) but I believe this would work - you should be able to add the unset commands into the VCL in the area where the URL blacklist settings are checked. So, change this:

```
if (req.url ~ "^(/index\.php/|/media/|/skin/|/js/)(?:(?:index|litespeed)\.php/)?admin") {
set req.backend = admin;
return (pipe);
}
```

to: 

```
if (req.url ~ "^(/index\.php/|/media/|/skin/|/js/)(?:(?:index|litespeed)\.php/)?admin") {
unset resp.http.X-Varnish;
unset resp.http.Via;
unset resp.http.X-Powered-By;
unset resp.http.Server;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;
set req.backend = admin;
return (pipe);
}
```

You could also do the same check in the custom vcl and it should work, once I update the VCL template files to put the custom VCL at the bottom.

**Comment by ADDISON74** [comment]
I will test both ways and come back with a feedback. In my opinion the best way is using the custom vcl which is already a feature in Turpentine. Please update the devel branch with VCL files having custom vcl placed at the end. I don't have any idea yet if there are any Varnish statements needed to be done in other places, not at the bottom.

**Comment by ADDISON74** [comment]
Inserting that code in VCL file it is not working. I get the following error:

```html
Message from VCC-compiler:
Unknown variable 'req.backend'
At: ('input' Line 167 Pos 5)
set req.backend = admin;
----###########---------

Running VCC-compiler failed, exited with 2

VCL compilation failed
Syntax check failed, not restarting ... failed!
```

**Comment by ADDISON74** [comment]
Second method is not working too. 

1) I edited app/code/community/Nexcessnet/Turpentine/misc/version-4.vcl and moved 

```html
## Custom VCL Logic

{{custom_vcl_include}}
```

at the very end of the file.

2) I created app/code/community/Nexcessnet/Turpentine/misc/custom_include.vcl with the following content:

```html
sub vcl_deliver {
unset resp.http.X-Varnish;
unset resp.http.Via;
unset resp.http.X-Powered-By;
unset resp.http.Server;
unset resp.http.X-Turpentine-Cache;
unset resp.http.X-Turpentine-Esi;
unset resp.http.X-Turpentine-Flush-Events;
unset resp.http.X-Turpentine-Block;
unset resp.http.X-Varnish-Session;
unset resp.http.X-Varnish-Host;
unset resp.http.X-Varnish-URL;
unset resp.http.X-Varnish-Set-Cookie;
}
```
Once I saved in backend Turpentine configuration I get at the end of default.vcl file the custom_include.vcl file content. Varnish restarts without any issues. Page I am visiting is not cached, included in URL Blacklist, but headers I wanted to hide are still there. Did I make something wrong?

**Comment by aricwatson** [comment]
I was able to make this work by creating the following custom include VCL file:

	sub vcl_recv {
		if (req.url ~ "{{url_base_regex}}(?:{{url_excludes}})") { 
			unset req.http.X-Varnish;
			unset req.http.Via;
			unset req.http.X-Powered-By;
			unset req.http.Server;
			unset req.http.X-Turpentine-Cache;
			unset req.http.X-Turpentine-Esi;
			unset req.http.X-Turpentine-Flush-Events;
			unset req.http.X-Turpentine-Block;
			unset req.http.X-Varnish-Session;
			unset req.http.X-Varnish-Host;
			unset req.http.X-Varnish-URL;
			unset req.http.X-Varnish-Set-Cookie;
			return (pipe);
		}
	}

This is using the most recent *devel* branch - where the custom_include.vcl file is included at the bottom of the generated VCL file.

Please test and see if it works for you!

**Comment by ADDISON74** [comment]
In my case it is not working. Something wrong in happening with custom_include.vcl file content. I pasted the same code as you pasted above. Here is what is placed by the end of default.vcl file. Both {{url_base_regex}} and {{url_excludes}} values are missing:

```html
sub vcl_recv {
if (req.url ~ "(?:)") {
unset req.http.X-Varnish;
unset req.http.Via;
unset req.http.X-Powered-By;
unset req.http.Server;
unset req.http.X-Turpentine-Cache;
unset req.http.X-Turpentine-Esi;
unset req.http.X-Turpentine-Flush-Events;
unset req.http.X-Turpentine-Block;
unset req.http.X-Varnish-Session;
unset req.http.X-Varnish-Host;
unset req.http.X-Varnish-URL;
unset req.http.X-Varnish-Set-Cookie;
return (pipe);
}
}
``` 

If I look  at the top of my default.vcl file I can see this:

```html
if (req.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?(?:admin|api|cron\.php|catalogsearch/term/popular/)" || req.url ~ "\?.*__from_store=") {
return (pipe);
}
```

I took this part of the code and pasted inside custom_include.vcl to see what's happening:

```html
sub vcl_recv {
if (req.url ~ "^(/media/|/skin/|/js/|/)(?:(?:index|litespeed)\.php/)?(?:admin|api|cron\.php|catalogsearch/term/popular/)" || req.url ~ "\?.*__from_store=") {
        unset req.http.X-Varnish;
        unset req.http.Via;
        unset req.http.X-Powered-By;
        unset req.http.Server;
        unset req.http.X-Turpentine-Cache;
        unset req.http.X-Turpentine-Esi;
        unset req.http.X-Turpentine-Flush-Events;
        unset req.http.X-Turpentine-Block;
        unset req.http.X-Varnish-Session;
        unset req.http.X-Varnish-Host;
        unset req.http.X-Varnish-URL;
        unset req.http.X-Varnish-Set-Cookie;
        return (pipe);
}
}
```

I saved the configuration in Magento Backend, I checked again defaul.vcl to see if the custom code is pasted correctly at the end and it is. Visiting one blacklist URL headers are still appearing which is not correct. 

In conclusion we have to understand why inside custom_include.vcl those 2 tags {{ ... }} are not getting the values according. Then why headers don't hide even unset tags are at the end of the file, out of the normal Turpentine configuration.

**Comment by aricwatson** [comment]
> Both {{url_base_regex}} and {{url_excludes}} values are missing

`{url_base_regex}` is based off your store's Base Url settings - what do you have set there for this site?

**Comment by aricwatson** [comment]
Closing for lack of feedback.

**Comment by ADDISON74** [comment]
Base Url in my case is http://www.mydomain.com and Base Secured Url https://www.mydomain.com. 

Your custom VCL file is not working in my case.
